tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp

```diff
@@ -73,8 +76,61 @@ uint32_t get_read_ptr(bool db_buf_switch) {
     return *reinterpret_cast<volatile tt_l1_ptr uint32_t*>(get_db_cb_rd_ptr_addr(db_buf_switch)) << 4;
 }
 
+inline __attribute__((always_inline)) volatile uint32_t* get_cq_completion_write_ptr() {
+    return reinterpret_cast<volatile uint32_t*>(CQ_COMPLETION_WRITE_PTR);
+}
+
+inline __attribute__((always_inline)) volatile uint32_t* get_cq_completion_read_ptr() {
+    return reinterpret_cast<volatile uint32_t*>(CQ_COMPLETION_READ_PTR);
+}
+
+FORCE_INLINE
+void cq_reserve_back(uint32_t data_size_B) {
+    DEBUG_STATUS('N', 'Q', 'R', 'B', 'W');
+    uint32_t data_size_16B = align(data_size_B, 32) >> 4;
+    uint32_t completion_rd_ptr_and_toggle;
+    uint32_t completion_rd_ptr;
+    uint32_t completion_rd_toggle;
+    do {
+        completion_rd_ptr_and_toggle = *get_cq_completion_read_ptr();
+        completion_rd_ptr = completion_rd_ptr_and_toggle & 0x7fffffff;
+        completion_rd_toggle = completion_rd_ptr_and_toggle >> 31;
+    } while (
+        cq_write_interface.completion_fifo_wr_ptr < completion_rd_ptr and cq_write_interface.completion_fifo_wr_ptr + data_size_16B > completion_rd_ptr or
+        (completion_rd_toggle != cq_write_interface.completion_fifo_wr_toggle and cq_write_interface.completion_fifo_wr_ptr == completion_rd_ptr)
+    );
+
+    DEBUG_STATUS('N', 'Q', 'R', 'B', 'D');
+}
+
+FORCE_INLINE
+void notify_host_of_cq_completion_write_pointer() {
+    constexpr static uint64_t pcie_address = (uint64_t(NOC_XY_ENCODING(PCIE_NOC_X, PCIE_NOC_Y)) << 32) | HOST_CQ_COMPLETION_WRITE_PTR;  // For now, we are writing to host hugepages at offset
+    uint32_t completion_wr_ptr_and_toggle = cq_write_interface.completion_fifo_wr_ptr | (cq_write_interface.completion_fifo_wr_toggle << 31);
+    // DPRINT << "device completion wr ptr " << cq_write_interface.completion_fifo_wr_ptr << ENDL();
+    volatile tt_l1_ptr uint32_t* completion_wr_ptr_addr = get_cq_completion_write_ptr();
+    completion_wr_ptr_addr[0] = completion_wr_ptr_and_toggle;
+    noc_async_write(CQ_COMPLETION_WRITE_PTR, pcie_address, 4);
+    noc_async_write_barrier();
+}
+
+FORCE_INLINE
+void cq_push_back(const uint32_t command_issue_region_size, uint32_t push_size_B) {
+    uint32_t push_size_16B = align(push_size_B, 32) >> 4;
+    cq_write_interface.completion_fifo_wr_ptr += push_size_16B;
+    if (cq_write_interface.completion_fifo_wr_ptr >= cq_write_interface.completion_fifo_limit) {
+        cq_write_interface.completion_fifo_wr_ptr = command_issue_region_size >> 4;
+        // Flip the toggle
+        cq_write_interface.completion_fifo_wr_toggle = not cq_write_interface.completion_fifo_wr_toggle;
+    }
+
+    // Notify host of updated completion wr ptr
+    notify_host_of_cq_completion_write_pointer();
+}
```

Comment:
Potentially we should now indicate that this is a "completion queue API"

Reply:
`notify_host_of_cq_completion_write_pointer` -> `notify_host_of_completion_queue_write_pointer`?